[KBFS-1100] Implement background work goroutine for JournalServer

akalin-keybase opened the pull request:
This background work goroutine auto-flushes journal entries
to the server. It will also do garbage collection in the future.

Change tlfJournalBundle from a POD type into the
tlfJournal type, which handles all the synchronization logic
with the block and MD journals, and the background work
goroutine.

Also add special files to pause/resume background work
goroutine.

Also add new and existing special files to dokan, which I
forgot to do previously.

akalin-keybase commented:
@jzila, giving this to you for review since @strib is out. It's good to get fresh eyes on this anyway -- let me know if you have any questions.

I'll also try to get some test coverage on this.

jzila commented on libfs/journal_control_file.go:
```diff
@@ -55,6 +63,12 @@ func (a JournalAction) Execute(
 		}
 
+	case JournalPause:
+		jServer.PauseAutoFlush(ctx, tlf)
```
Do writes block while a flush is happening?

akalin-keybase replied:
Yeap, but that's not intended: see https://keybase.atlassian.net/browse/KBFS-1433

jzila commented on libkbfs/journal_server.go:
```diff
@@ -176,18 +218,130 @@ func (j *JournalServer) Enable(ctx context.Context, tlfID TlfID) (err error) {
+	}
+
+	select {
+	case bundle.pauseCh <- struct{}{}:
```
If it's currently Flushing then this pause will be lost.

akalin-keybase replied:
This is fine, I think, since the channel is buffered for one item. That just means that multiple pause requests (or resume requests) in a row will get coalesced into one, which is fine, since those operations are just meant for ad hoc testing. Specifically, if you do a pause, resume, pause in quick succession, the second pause might be dropped, but I think that's fine. Although come to think of it, if it's in the middle of flushing a long list of requests, it should stop in the middle instead of waiting for the whole list. I'll do that.

I'll also add some comments and change some names to make it clearer what's going on.

jzila replied:
Makes sense, thanks!

jzila commented on libkbfs/journal_server.go:
```diff
@@ -301,6 +455,11 @@ func (j *JournalServer) Disable(ctx context.Context, tlfID TlfID) (err error) {
 	}
 
+	select {
+	case bundle.shutdownCh <- struct{}{}:
```
This should be a `close()` IMO, since you don't want the shutdown write to get lost.

akalin-keybase replied:
For the same reason as above, multiple shutdown calls get coalesced into one, which is fine. If I close here, then that's problematic since subsequent closes will panic.

jzila commented on libkbfs/journal_server.go:
```diff
@@ -361,3 +520,15 @@ func (j *JournalServer) JournalStatus(tlfID TlfID) (TLFJournalStatus, error) {
+	defer j.lock.Unlock()
+	for _, bundle := range j.tlfBundles {
+		select {
+		case bundle.shutdownCh <- struct{}{}:
```
This should be a close(), sinc eyou don't want the shutdown write to get lost.

akalin-keybase replied:
Ditto.

jzila commented on libkbfs/journal_server.go:
```diff
@@ -176,18 +218,130 @@ func (j *JournalServer) Enable(ctx context.Context, tlfID TlfID) (err error) {
 	return nil
 }
 
+func (j *JournalServer) autoFlush(
```
It seems to me like this operates entirely on fields of `tlfJournalBundle`. It might be cleaner to have this method's receiver be `tlfJournalBundle`.

akalin-keybase replied:
Done!

jzila commented on libkbfs/journal_server.go:
```diff
@@ -176,18 +218,130 @@ func (j *JournalServer) Enable(ctx context.Context, tlfID TlfID) (err error) {
+	}
+
+	select {
+	case bundle.hasWorkCh <- struct{}{}:
```
If a flush is currently happening, this notification will get lost.

akalin-keybase replied:
Yeap, that's fine for the same reason as above.

jzila commented:
First pass done. Much love!

jzila commented on libkbfs/journal_server.go:
```diff
@@ -176,18 +218,130 @@ func (j *JournalServer) Enable(ctx context.Context, tlfID TlfID) (err error) {
+}
+
 // Flush flushes the write journal for the given TLF.
 func (j *JournalServer) Flush(ctx context.Context, tlfID TlfID) (err error) {
```
I think this method should be delegated to `tlfJournalBundle` with `j`'s `Config` passed in.

akalin-keybase replied:
Done!

jzila commented:
K. LGTM after feedback implemented, assuming tests come later.

akalin-keybase commented:
I implemented your feedback, but that ended up causing me to convert tlfJournalBundle from a POD type into a "smart" type with synchronization logic, etc. I think this was a substantial enough change that you should take another look.

I also added some tests! PTAL, thanks.

akalin-keybase commented:
Looks like I have a test flake to track down.

jzila commented:
@akalin-keybase might just be that we need to increase the timeout for libkbfs tests.

akalin-keybase commented:
Okay, rebased and fixed test failure. Also fixed up the tests to individually timeout (which as a side effect produces actually-useful test output, since kill -ABRT and the test timeout doesn't print out test output...urgh.)

jzila commented on libkbfs/journal_block_server.go:
```diff
@@ -17,41 +17,26 @@ var _ BlockServer = journalBlockServer{}
+	if tlfJournal, ok := j.jServer.getTLFJournal(tlfID); ok {
+		data, serverHalf, err := tlfJournal.getBlockDataWithContext(
+			id, context)
+		if _, ok := err.(BServerErrorBlockNonExistent); ok {
```
You're shadowing `ok` here, which makes reading a little confusing. Maybe:
```go
switch err.(type) {
case nil:
    return data, serverHalf, nil
case BServerErrorBlockNonExistent:
    return j.BlockServer.Get(ctx, tlfID, id, context)
default:
    return nil, BlockCryptKeyServerHalf{}, err
}
```

akalin-keybase replied:
Done.

jzila commented on libkbfs/journal_md_ops.go:
```diff
@@ -42,25 +42,12 @@ func (j journalMDOps) getHeadFromJournal(
-	bundle, ok := j.jServer.getBundle(id)
+	tlfJournal, ok := j.jServer.getTLFJournal(id)
 	if !ok {
 		return ImmutableRootMetadata{}, nil
```
shouldn't this be an error?

akalin-keybase replied:
No -- journalMDOps overlays the MDs from the journal on top of the 'regular' MDOps (talking to the MDServer), so if there's no journal at all for a particular TLF, then that just means that it falls back to the server.

jzila commented on libkbfs/tlf_journal.go:
```diff
@@ -0,0 +1,611 @@
+	var errCh <-chan error
+	var bwCancel context.CancelFunc
+	for {
+		switch {
```
you can collapse this to a `switch bws`, and branch on `errCh` separately in the first case below.

jzila replied:
Side note, this is an elegant state machine :)

akalin-keybase replied:
I thought about that, but I'd rather have all the states be on the same indent level. Plus, it's already so indented. :(

jzila commented on libkbfs/tlf_journal.go:
```diff
@@ -0,0 +1,611 @@
+	// TODO: Parallelize block puts.
+
+	for {
+		flushed, err := j.flushOneBlockOp(ctx)
```
Seems like pausing won't actually pause this.

jzila replied:
Also seems like this workload could continue even after shutdown.

akalin-keybase replied:
Pausing and shutting down will cause the worker goroutine to exit -- we pass in bwCtx to doBackgroundWork (which passes it to the worker goroutine), and we call the corresponding bwCancel function when we exit out of the busy state. There are some tests that exercise this, and without that, they would hang.

I added some comments to call this out a bit better.

jzila commented on libkbfs/tlf_journal.go:
```diff
@@ -0,0 +1,611 @@
+		return err
+	}
+
+	select {
```
what happened to `j.signalWork()`? This `select` appears many times.

akalin-keybase replied:
Ah, added it back in.

jzila commented on libkbfs/tlf_journal_test.go:
```diff
@@ -0,0 +1,227 @@
+	bs.waitForPut(ctx, t)
+	delegate.requireNextState(ctx, t, bwBusy)
+
+	// Should still be able to shut down while busy.
```
true, but it will leave behind a goroutine.

akalin-keybase replied:
It won't, I think -- the teardown process calls cancel(), which cancels the main test context, which causes the hanging block/md server to stop hanging and return.

I wish there were a way to check goroutine counts in tests, but alas...

jzila commented on libkbfs/journal_md_ops.go:
```diff
@@ -345,22 +319,10 @@ func (j journalMDOps) GetUnmergedRange(
 func (j journalMDOps) Put(ctx context.Context, rmd *RootMetadata) (
 	MdID, error) {
-	bundle, ok := j.jServer.getBundle(rmd.TlfID())
+	tlfJournal, ok := j.jServer.getTLFJournal(rmd.TlfID())
```
Same name as type, but that's a rule we break everywhere so feel free to leave it if you want to.

akalin-keybase replied:
Yeah, I can't think of a better name.

jzila commented:
Looks great, I like the workload being in `tlfJournal`. Some comments, mostly formatting.

jzila commented:
I'd say that you can merge once the comments are addressed.

jzila commented:
LGTM once tests pass.
